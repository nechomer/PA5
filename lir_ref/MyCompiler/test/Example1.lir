re_null_ref_str:    "Runtime Error: Null pointer dereference!"
re_array_index_str: "Runtime Error: Array index out of bounds!"
re_array_alloc_str: "Runtime Error: Array allocation with non-positive array size!"
re_zero_div_str:    "Runtime Error: Division by zero!"
str1: "str_25"
str2: "hi"
str3: "Refaeli"

# class A
# Dispatch vector:
_DV_A: [_A_bar]
# Field offsets
# 1: str
# 2: i

# class B
# Dispatch vector:
_DV_B: [_A_bar, _B_vfunc]
# Field offsets
# 1: str
# 2: i
# 3: j

# class C
# Dispatch vector:
_DV_C: []
# Field offsets


__checkNullRef:
	# a is a reference
	Move a, R0
	Compare 0, R0
	JumpTrue _rc_error_label1
	Return Rdummy
_rc_error_label1:
	Library __println(re_null_ref_str),Rdummy
	Library __exit(1),Rdummy

__checkArrayAccess:
	# a is an arry
	# i is an index register
	ArrayLength a, R0
	Compare i, R0
	JumpLE _rc_error_label2
	Move i, R0
	Compare 0, R0
	JumpL _rc_error_label2
	Return Rdummy
_rc_error_label2:
	Library __println(re_array_index_str),Rdummy
	Library __exit(1),Rdummy

__checkSize:
	# n is an array length
	Move n, R0
	Compare 0, R0
	JumpLE _rc_error_label3
	Return Rdummy
_rc_error_label3:
	Library __println(re_array_alloc_str),Rdummy
	Library __exit(1),Rdummy

__checkZero:
	# x is a value that should no be zero
	Move x, R0
	Compare 0, R0
	JumpTrue _rc_error_label4
	Return Rdummy
_rc_error_label4:
	Library __println(re_zero_div_str),Rdummy
	Library __exit(1),Rdummy

_A_zfunc:
Move 5, R0
Move R0, str_26
Library __allocateObject(12), R0
MoveField _DV_A, R0.0
Move R0, a_27
Move 6, R0
Move R0, str_26
Move str1, R0
Move a_27, R1
StaticCall __checkNullRef(a=R1),Rdummy
MoveField R0, R1.1
Move a_27, R1
StaticCall __checkNullRef(a=R1),Rdummy
MoveField R1.1, R0
Move R0, str_28
Move str_28, R1
Library __println(R1), Rdummy
Move str2, R1
StaticCall _A_foo(str_25_29=R1), Rdummy
Move str_26, R1
Library __printi(R1), Rdummy
Move a_27, R1
StaticCall __checkNullRef(a=R1),Rdummy
Move str3, R2
VirtualCall R1.0(str_28_31=R2), Rdummy
Move a_27, R2
StaticCall __checkNullRef(a=R2),Rdummy
MoveField R2.1, R1
Library __println(R1), Rdummy
Move 0, R0
Return R0

_A_foo:
Move 5, R0
Move R0, i_30
Move str_25_29, R1
Library __println(R1), Rdummy
Return Rdummy

_A_bar:
Move str_28_31, R0
Move this, R1
MoveField R0, R1.1
Return Rdummy

_B_vfunc:
Move 0, R0
Move a_32, R1
Move this, R2
Compare R1, R2
JumpFalse _logical_op_end_3
Move 1, R0
_logical_op_end_3:
Compare 0, R0
JumpTrue _false_label_1
Return Rdummy
_false_label_1:
Move 0, R0
Move this, R2
MoveField R2.2, R1
Move this, R3
StaticCall __checkNullRef(a=R3),Rdummy
MoveField R3.3, R2
Compare R1, R2
JumpLE _logical_op_end_6
Move 1, R0
_logical_op_end_6:
Compare 0, R0
JumpTrue _false_label_4
Move this, R2
MoveField R2.1, R1
Library __println(R1), Rdummy
_false_label_4:
Move 0, R0
Move a_32, R2
StaticCall __checkNullRef(a=R2),Rdummy
MoveField R2.2, R1
Move this, R3
MoveField R3.3, R2
Compare R1, R2
JumpLE _logical_op_end_9
Move 1, R0
_logical_op_end_9:
Compare 0, R0
JumpTrue _false_label_7
Move a_32, R2
StaticCall __checkNullRef(a=R2),Rdummy
MoveField R2.1, R1
Library __println(R1), Rdummy
_false_label_7:
Return Rdummy

_B_zfunc:
Move 0, R0
Return R0

_B_main:
Return Rdummy


_ic_main:
Move 1, R1
StaticCall _A_zfunc(b_25=R1), R0
Library __exit(0),Rdummy
