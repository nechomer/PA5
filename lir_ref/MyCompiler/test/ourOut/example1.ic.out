Parsed ..\test\libic.sig successfully!
Parsed ..\test\LegalPrograms\example1.ic successfully!

Global Symbol Table: ..\test\LegalPrograms\example1.ic
	Class: A
	Class: B
	Class: C
Children tables: A, C, Library

Class Symbol Table: A
	Field: string str
	Field: int i
	Static method: sfunc {boolean -> int}
	Virtual method: vfunc {A, int, int -> void}
Children tables: B, sfunc, vfunc

Class Symbol Table: C
	Static method: main {string[] -> void}
Children tables: main

Class Symbol Table: Library
	Static method: println {string -> void}
	Static method: print {string -> void}
	Static method: printi {int -> void}
	Static method: printb {boolean -> void}
	Static method: readi { -> int}
	Static method: readln { -> string}
	Static method: eof { -> boolean}
	Static method: stoi {string, int -> int}
	Static method: itos {int -> string}
	Static method: stoa {string -> int[]}
	Static method: atos {int[] -> string}
	Static method: random {int -> int}
	Static method: time { -> int}
	Static method: exit {int -> int}

Class Symbol Table: B
	Field: int j
	Virtual method: vfunc {A, int, int -> void}
Children tables: vfunc

Method Symbol Table: sfunc
	Parameter: boolean b
Children tables: statement block in sfunc, statement block in sfunc

Method Symbol Table: vfunc
	Parameter: A a
	Parameter: int max
	Parameter: int val
	Local variable: int[] arr
	Local variable: int i
Children tables: statement block in vfunc

Method Symbol Table: vfunc
	Parameter: A a
	Parameter: int max
	Parameter: int val

Statement Block Symbol Table ( located in sfunc )
Children tables: statement block in statement block in sfunc

Statement Block Symbol Table ( located in sfunc )
	Local variable: A x

Statement Block Symbol Table ( located in statement block in sfunc )
	Local variable: int k

Statement Block Symbol Table ( located in vfunc )

Method Symbol Table: main
	Parameter: string[] args
	Local variable: A[][] a

Type Table: ..\test\LegalPrograms\example1.ic
	1: Primitive type: int
	0: Primitive type: boolean
	4: Primitive type: null
	2: Primitive type: string
	3: Primitive type: void
	5: Class: A
	6: Class: B, Superclass ID: 5
	7: Class: C
	11: Array type: string[]
	22: Array type: int[]
	29: Array type: int
	32: Array type: A[]
	33: Array type: A[][]
	8: Method type: {boolean -> int}
	9: Method type: {A, int, int -> void}
	12: Method type: {string[] -> void}
	13: Method type: {string -> void}
	15: Method type: {int -> void}
	16: Method type: {boolean -> void}
	17: Method type: { -> int}
	18: Method type: { -> string}
	19: Method type: { -> boolean}
	20: Method type: {string, int -> int}
	21: Method type: {int -> string}
	23: Method type: {string -> int[]}
	25: Method type: {int[] -> string}
	26: Method type: {int -> int}

